core/arch/arm/kernel/secure_partition.c

```diff
@@ -603,6 +603,60 @@ static TEE_Result handle_fdt_dev_regions(struct sp_ctx *ctx, void *fdt)
 	return TEE_SUCCESS;
 }
 
+static TEE_Result read_manifest_endpoint_id(struct sp_session *s,
+					    const void *fdt)
+{
+	uint32_t endpoint_id = 0;
+
+	/* The endpoint id can be optionally defined in the manifest file. We
+	 * have to map the id inside the manifest to the SP if it's defined.
+	 * If not, the endpoint id generated inside new_session_id() will be
+	 * used.
+	 */
+	if (!sp_dt_get_u32(fdt, 0, "id", &endpoint_id)) {
+		struct sp_session *a_session = sp_get_session(endpoint_id);
+
+		if (endpoint_id <= SPMC_ENDPOINT_ID)
+			return TEE_ERROR_BAD_FORMAT;
+		/*
+		 * We don't know in which order the SPs are loaded. It can
+		 * be that the endpoint id defined in the manifest could already
+		 * be generated by new_session_id() and used by another SP. If
+		 * this is the case, we swap the ID's of the two SPs.
+		 */
+		if (a_session)
+			a_session->endpoint_id = s->endpoint_id;
```

Comment:
This shouldn't be permitted if the endpoint ID has already been assigned from a manifest. Doing the double-check later in `check_endpoint_ids()` seems a bit backward.

Reply:
Ack

Reply:
I rewrote the patch a bit to do the check before assigning the endpoint id to the SP.